sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -253,29 +254,38 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
       val dsOptions = new DataSourceOptions(options.asJava)
       provider.getTable(dsOptions) match {
         case table: SupportsBatchWrite =>
-          if (mode == SaveMode.Append) {
-            val relation = DataSourceV2Relation.create(table, options)
-            runCommand(df.sparkSession, "save") {
-              AppendData.byName(relation, df.logicalPlan)
-            }
-          } else {
-            val writeBuilder = table.newWriteBuilder(dsOptions)
-              .withQueryId(UUID.randomUUID().toString)
-              .withInputDataSchema(df.logicalPlan.schema)
-            writeBuilder match {
-              case s: SupportsSaveMode =>
-                val write = s.mode(mode).buildForBatch()
-                // It can only return null with `SupportsSaveMode`. We can clean it up after
-                // removing `SupportsSaveMode`.
-                if (write != null) {
-                  runCommand(df.sparkSession, "save") {
-                    WriteToDataSourceV2(write, df.logicalPlan)
+          lazy val relation = DataSourceV2Relation.create(table, options)
+          mode match {
+            case SaveMode.Append =>
+              runCommand(df.sparkSession, "save") {
+                AppendData.byName(relation, df.logicalPlan)
+              }
+
+            case SaveMode.Overwrite =>
+              // truncate the table
+              runCommand(df.sparkSession, "save") {
+                OverwriteByExpression.byName(relation, df.logicalPlan, Literal(true))
```

Comment:
@gatorsmile, I verified that file system sources truncate in all cases when mode is overwrite. I've updated this to always truncate as you requested.

Reply:
-> `filter = Literal(true)`.

Reply:
how shall we indicate a truncate? using true literal as the overwrite expression works, another option is to use `Nil`. No overwrite expressions can also mean truncating.

Reply:
@cloud-fan, I think that `Literal(true)` is better than using `Nil`. This translates directly to an `AlwaysTrue` filter, which has obvious behavior: `DELETE FROM table WHERE true`.

Using `Nil` or no filters to signal that all rows should be deleted is not obvious and requires more documentation for the contract with implementations.

Reply:
Actually we already have this contract. e.g. `INSERT OVERWRITE tbl SELECT ...` without WHERE will truncate the table and write new data. If the contract is here for end-users, I think it's OK for data source developers as well.

Anyway I think this shouldn't block this PR and we can discuss it later.